src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetPackageSearchService.cs

```diff
@@ -170,9 +176,47 @@ public NuGetPackageSearchService(ServiceActivationOptions options, IServiceBroke
 
             IPackageMetadataProvider packageMetadataProvider = await GetPackageMetadataProviderAsync(packageSources, cancellationToken);
             IPackageSearchMetadata packageMetadata = await packageMetadataProvider.GetPackageMetadataAsync(identity, includePrerelease, cancellationToken);
-            IEnumerable<VersionInfo> versions = await packageMetadata.GetVersionsAsync();
 
-            return await Task.WhenAll(versions.Select(v => VersionInfoContextInfo.CreateAsync(v).AsTask()));
+            var cacheEntry = new PackageSearchMetadataCacheItem(packageMetadata, packageMetadataProvider);
+            cacheEntry.UpdateSearchMetadata(packageMetadata);
+
+            object cacheObject = PackageSearchMetadataMemoryCache.AddOrGetExisting(cacheId, cacheEntry, _cacheItemPolicy);
+
+            return await ((PackageSearchMetadataCacheItem)cacheObject).AllVersionsContextInfo;
+        }
+
+        public async ValueTask<IReadOnlyCollection<VersionInfoContextInfo>> GetPackageVersionsAsync(
```

Comment:
I am looking at this code path for almost first time, would it be possible to add some comments so that I can understand the code flow a bit better.